modules/relayers/src/extension/priority.rs

```diff
@@ -22,31 +22,89 @@
 //! single message with nonce `N`, then the transaction with nonces `N..=N+100` will
 //! be rejected. This can lower bridge throughput down to one message per block.
 
-use bp_messages::MessageNonce;
+use crate::{Config as RelayersConfig, Pallet as RelayersPallet};
+
+use bp_messages::{LaneId, MessageNonce};
 use frame_support::traits::Get;
-use sp_runtime::transaction_validity::TransactionPriority;
+use frame_system::{pallet_prelude::BlockNumberFor, Pallet as SystemPallet};
+use sp_runtime::{
+	traits::{One, Zero},
+	transaction_validity::TransactionPriority,
+};
 
 // reexport everything from `integrity_tests` module
 pub use integrity_tests::*;
 
-/// Compute priority boost for message delivery transaction that delivers
-/// given number of messages.
-pub fn compute_priority_boost<PriorityBoostPerMessage>(
+/// Compute total priority boost for message delivery transaction.
+pub fn compute_priority_boost<R: RelayersConfig>(
+	lane_id: LaneId,
 	messages: MessageNonce,
+	relayer: &R::AccountId,
 ) -> TransactionPriority
 where
-	PriorityBoostPerMessage: Get<TransactionPriority>,
+	usize: TryFrom<BlockNumberFor<R>>,
 {
+	compute_per_message_priority_boost::<R::PriorityBoostPerMessage>(messages)
+		.saturating_add(compute_per_lane_priority_boost::<R>(lane_id, relayer))
+}
+
+/// Compute priority boost for message delivery transaction, that depends on number
+/// of bundled messages.
+fn compute_per_message_priority_boost<PriorityBoostPerMessage: Get<TransactionPriority>>(
+	messages: MessageNonce,
+) -> TransactionPriority {
 	// we don't want any boost for transaction with single message => minus one
 	PriorityBoostPerMessage::get().saturating_mul(messages.saturating_sub(1))
 }
 
+/// Compute priority boost for message delivery transaction, that depends on
```

Comment:
Another important (and fragile) thing here is that if validator own collator(s), he may simply ignore priorities and include affilated relayer transactions into blocks that he authors. So honest relayer may not receive rewards even if he is submitting proper transactions in assigned slots. Normally we assume that the collators set is honest and decentralized (so it is solved elsewhere), but here we build the important subsystem (relayer rewards) on top of that. 

Options I see:
- ignore that fact. If we admit the possibility of centralized collators set, then we also admit the possibility of censoring particular transactions - e.g. such collator set may simply halt one of bridges (lanes) by ignoring delivery transactions on those lanes. So still we need to solve it elsewhere;
- have some kind of misbehavior reports. But proving that the honest relayer has submitted valid transaction, that was not "mined" is impossible. All other solutions (e.g. only accepting delivery transactions from relayer that has submitted finality proof for next `N` blocks) still allow affilated collators to censorship such transactions and get rewarded;
- have a stricter laws. I.e. instead of bumping priority for relayer at the head of the queue, we simply reject all other relayer transactions during his slot. This won't protect us from censorship anyway.

So I suggest us to go with the option (1) - I don't think that we can (and should) solve it here. So just wanted to share and warn that this ^^^ is possible.

Reply:
I am not very familiar with collators. How would this work exactly ? How many collators would the relayer need to own in order to bypass our priorities mechanism ? Is there any way for the validators to realize that these collators are assigning wrong priorities to the bridge-related transactions and avoid these collators ? I'm wondering if it wouldn't be more advantageous for a relayer to run a set of modified collators instead of registering as a relayer, because:
- registering as a relayer requires staking, which is a disadvantage
- running a set of collators could lead to more mined transactions maybe ? So more rewards